### PR TITLE
fix: generate page_break for skipped pages in export functions

### LINKED ISSUE #472: Export functions fail to generate <page_break> tags for non-consecutive (skipped) pages
## Description 

When parsing PDF documents, if specific pages fail to parse (e.g., due to exceptions caught in the pipeline) and are excluded from the `doc.pages` list, the export functions (`export_to_doctags`, `export_to_html`, `export_to_markdown`) do not generate `page break` markers for these gaps. This causes page count mismatch issues in downstream processing. 

### Attachments
[test.pdf](https://github.com/user-attachments/files/24523289/test.pdf)
* `test.pdf` - A minimal reproduction file derived from a larger document. The content has been intentionally corrupted for confidentiality, but the parsing error still reproduces.
  * Structure: Consists of 4 pages (Pages 78, 79, 83, 84).
  * Behavior: Parsing succeeds for pages 78 and 84, but fails for pages 79 and 83. 

1.  Convert the attached `test.pdf` using `DocumentConverter`.
2.  Export using `export_to_doctags()`, `export_to_markdown(page_break_placeholder=str)`, `export_to_html(split_page_view=True)`.
3.  Count the `page break` 

**Reproduction Script:** 

```python
converter = DocumentConverter(
    format_options={
        InputFormat.PDF: PdfFormatOption(
            pipeline_options=pipeline_options,
        ),
    },
) 

doc = converter.convert(source).document

doctags_output = doc.export_to_doctags()
markdown_output = doc.export_to_markdown(page_break_placeholder="===PAGE_BREAK===")
html_output = doc.export_to_html(split_page_view=True) 

print(doctags_output)
print(markdown_output)
print(html_output)
``` 

## Expected Behavior
* **Export:** export fuction should generate `pagebreak` tags for **all pages**, including failed ones, so that page numbering remains consistent. 

## Actual Behavior
**Missing `page break` for failed pages:**
If pages 1, 2, 4, 5 succeed but page 3 fails, the output currently looks like this: 

```xml
... content ...      1page
<page_break>
... content ...      2page
<page_break>
... content ...      4page
<page_break>
... content ...      5page
``` 

```markdown
... content ...      1page
===PAGE_BREAK===
... content ...      2page
===PAGE_BREAK===
... content ...      4page
===PAGE_BREAK===
... content ...      5page
```

```html
<td>
<div class="page">
... content ...      1page
</div>
</td> 

<td>
<div class="page">
... content ...      2page
</div>
</td> 

<td>
<div class="page">
... content ...      4page
</div>
</td> 

<td>
<div class="page">
... content ...      5page
</div>
</td>
```

### Environment 

| Component | Version / Details |
| :--- | :--- |
| **docling version** | 2.31.1 |
| **docling-core version** | 2.31.0 |
| **Python** | 3.11 |
| **OS** | macOS |

### PR DESCRIPTION
**Pull request by jhchoi1182**
## Description

Currently, pages that fail during PDF parsing are silently skipped. Taking the test.pdf attached to the issue (composed of pages 78, 79, 83, and 84) as an example, pages 79 and 83 failed to parse, causing page 78 to be followed immediately by page 84. Previously, this resulted in the generation of only a single `<page_break>` for page 84. This made it impossible for users to detect that pages 79 and 83 were missing, leading to potential issues during post-processing.

This PR modifies the `_iterate_items()` function to generate individual `<page_break>` markers for each skipped page when parsing failures result in non-consecutive page transitions. For instance, when jumping from page 78 to 84, the function now generates page breaks for the missing pages (79, 83) in addition to page 84, ensuring all skipped pages are explicitly marked.

### Changes

**common.py:**
- Added `_PageBreakNode` class to represent page break nodes with `prev_page` and `next_page` fields
- Added `_yield_page_breaks()` helper function that generates page break nodes for each page in a non-consecutive range
- Updated `_iterate_items()` to use the new helper, ensuring all skipped pages get their own `<page_break>` marker

**html.py:**
- Updated `serialize_doc()` to handle skipped pages in `SPLIT_PAGE` mode
- Skipped pages are now rendered as empty rows in the split page view, maintaining page numbering consistency
- Added logic to track all physical pages from page breaks and render the full range including missing pages

### Affected Export Functions

| Function | Behavior |
|----------|----------|
| `export_to_doctags()` | Generates `<page_break>` tokens for skipped pages |
| `export_to_markdown(page_break_placeholder=str)` | Generates placeholder for skipped pages |
| `export_to_html(split_page_view=True)` | Renders skipped pages as empty rows in split view |

### Design Decision

When page breaks are enabled, breaks are now generated for all page transitions, including pages that failed to parse. This is the default behavior without additional parameters, as it provides the most intuitive result: users can detect missing pages in the output. If maintainers prefer to make this configurable, a `skip_missing_page_breaks` parameter could be added to `CommonParams` in a follow-up PR.

### Example

**`export_to_doctags()`**

Before (page 78, 79, 83, 84):
```xml
<doctag>
... content ...      78page
<page_break>
... content ...      84page
</doctag>
```
After (page 78, 79, 83, 84):
```xml
<doctag>
... content ...      78page
<page_break>
<page_break>  (79, 83page)
<page_break>
... content ...      84page
</doctag>
```

**`export_to_markdown(page_break_placeholder="---PAGE BREAK---")`**

Before (page 78, 79, 83, 84):
```markdown
... content ...      78page

---PAGE BREAK---

... content ...      84page
```
After (page 78, 79, 83, 84):
```markdown
... content ...      78page

---PAGE BREAK---

---PAGE BREAK---

---PAGE BREAK---

... content ...      84page
```

**`export_to_html(split_page_view=True)`**

Before (page 78, 79, 83, 84):
```html
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
... content ...      78page

</div>
</td>
</tr>
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
... content ...      84page

</div>
</td>
</tr>
```

After (page 78, 79, 83, 84):
```html
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
... content ...      78page

</div>
</td>
</tr>
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
                     (79page)

</div>
</td>
</tr>
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
                     (83page)

</div>
</td>
</tr>
<tr>
<td>
<figure>no page-image found</figure>
</td>
<td>
<div class='page'>
... content ...      84page

</div>
</td>
</tr>
```

## Testing
Unit tests for this feature are available in `test/test_page_break_skipped_pages.py`. Run the tests with:

```bash
pytest test/test_page_break_skipped_pages.py -v
```

## Issue resolved by this Pull Request:
Resolves #472

## Checklist:
- [x]  My code follows the style guidelines of this project
- [x]  I have performed a self-review of my own code
- [x]  New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docling_core/transforms/serializer/common.py [hunk 1]
```diff
@@ -78,6 +78,36 @@ class _PageBreakSerResult(SerializationResult):
...
+
+    Args:
+        prev_page: The last seen page number.
+        next_page: The current page number.
+        lvl: The nesting level for the yielded nodes.
+        start_index: The starting index for page break node IDs.
+
+    Yields:
+        Tuples of (PageBreakNode, level, next_index) for each page transition.
+    """
```
[the remark this answers is not in the document]
Comment (jhchoi1182) [reply docling_core/transforms/serializer/common.py:109]: Thanks for the feedback! Updated the docstring to use physical page indices (1->4 yields 3 breaks) instead of PDF labels, which better reflects actual runtime behavior.

File: docling_core/transforms/serializer/common.py [hunk 2]
```diff
@@ -113,22 +143,19 @@ def _iterate_items(
...
                         if prev_page_nr is not None and page_no > prev_page_nr:
-                            yield _PageBreakNode(
-                                self_ref=f"#/pb/{page_break_i}",
-                                prev_page=prev_page_nr,
-                                next_page=page_no,
-                            ), lvl
+                            for pb_node, pb_lvl, page_break_i in _yield_page_breaks(
+                                prev_page_nr, page_no, lvl, page_break_i
+                            ):
+                                yield pb_node, pb_lvl
```
[the remark this answers is not in the document]
Comment (jhchoi1182) [reply docling_core/transforms/serializer/common.py:165]: Good catch! You're right that this could cause duplicate page breaks.

I've applied the fix with a slight modification - moving `prev_page_nr = page_no` outside the `if` block to ensure it's always updated when we find the first DocItem of a group, regardless of whether a page break was generated. This handles edge cases like:
- When `prev_page_nr` is `None` (first item in document)
- When the group starts on the same page (no page break needed, but still need to track the page)

Thanks for the review!

File: docling_core/transforms/serializer/common.py [hunk 3]
```diff
@@ -78,6 +78,36 @@ class _PageBreakSerResult(SerializationResult):
...
+        Tuples of (PageBreakNode, level, next_index) for each page transition.
+    """
+    idx = start_index
+    for page in range(prev_page + 1, next_page + 1):
+        yield _PageBreakNode(
+            self_ref=f"#/pb/{idx}",
+            prev_page=page - 1,
+            next_page=page,
+        ), lvl, idx + 1
+        idx += 1
```
[the remark this answers is not in the document]
Comment (jhchoi1182) [reply docling_core/transforms/serializer/common.py:124]: Added test coverage in `test/test_page_break_skipped_pages.py` addressing all suggested items:

1. **Multiple page breaks for skipped pages**: 
   - `test_doctags_page_breaks_skipped_2pages`: verifies 3 page breaks when pages 2,3 are missing
   - `test_markdown_page_breaks_skipped_2pages`: same for markdown export

2. **page_break_i index maintenance**: 
   - Verified indirectly through correct page break counts (if index were incorrect, counts would be wrong)

3. **prev_page/next_page values**: 
   - These are internal implementation details not exposed in export formats
   - Correctness verified indirectly via page break count validation

4. **Edge cases (1 page vs multiple pages)**:
   - `skipped_1page.json`: pages [1, 3] - tests single page skip (1->2->3)
   - `skipped_2pages.json`: pages [1, 4] - tests multiple page skip (1->2->3->4)

All 9 tests pass.

**Comment by jhchoi1182** [comment]
I found that my previous PR code conflicted with the `.filter()` method.

To resolve this, I've submitted a separate PR to the docling repository that adds code to track failed pages by adding them to `DoclingDocument.pages`. This companion PR in docling-core modifies `_yield_page_breaks()` to only generate page breaks for pages present in the `pages` dict.

### How it works:

* **docling:** `_add_failed_pages_to_document()` method adds failed/skipped pages to `DoclingDocument.pages` with their size information (empty content)
* **docling-core:** `_yield_page_breaks()` now takes `page_numbers` parameter (extracted from `doc.pages.keys()`) and only generates page breaks for pages in that set

### This approach ensures:

* ✅ Failed pages get proper page break markers (pages are in `pages` dict)
* ✅ `filter()` method continues to work correctly (excluded pages are removed from `pages` dict, so no spurious page breaks)
* ✅ Filter + failed page scenario works (e.g., filtering `{2,3,5}` with page 3 failing still generates correct breaks)

Related PR: https://github.com/docling-project/docling/pull/2939

**Comment by serboor** [comment]
We hit this in production and would like to add two data points that may help move this forward.

**1. The failure mode is wider than parsing failures.** We see missing page breaks with a
`ConversionStatus.SUCCESS` conversion, 0 errors and every page present in `document.pages`. The
trigger is simply a page with **no items at all**: a fully scanned 32-page PDF where the layout model
only produced items on 18 pages emitted **17 breaks instead of 31**. Since our OCR injection matches
markdown sections to pages positionally, 14 pages of already-computed OCR were silently dropped. The
fix in this PR (one break per skipped page) covers this case as well, which we think strengthens it:
it is not only about parsing errors.

**2. The page-by-page workaround is exact but 150× more expensive.** Exporting each page with
`export_to_markdown(page_no=n)` and joining with the placeholder reproduces the corrected output
byte-for-byte, so it is a good oracle for tests — but the cost grows with pages × items and is
unusable in a pipeline: on our 287-page document it takes **179.54 s** versus **1.06 s** for a single
global export (154-page document: 33.47 s vs 0.64 s; 63-page: 8.85 s vs 0.20 s). That is the reason
we cannot simply work around the issue and would rather see it fixed in `_iterate_items`.

**3. There is a second, independent placement bug in the same block**, which this PR does not cover:
a page break for a `GroupItem` whose children start on a new page is emitted *after* the group.
Reported separately with a self-contained repro in https://github.com/docling-project/docling/issues/3920. A fix for both could share the same
loop.

Environment: docling 2.115.0, docling-core 2.87.1, Python 3.10.12.
